### DISCUSSION TITLE
Why does Diesel require custom member types to implement `Iterator`?

### DISCUSSION THREAD

**Discussion by McDic** [diesel-rs/diesel#4038] [Q&A] [answered]
I was coding custom proxy types on serialization and de-serializations, following [this discussion](https://github.com/diesel-rs/diesel/discussions/3835).

Let's suppose following codes.

```rust
use chrono::Duration;
use diesel::pg::data_types::PgInterval;
use diesel::{Insertable, Queryable, Selectable};

mod schema {
    diesel::table! {
        fantastic (id) {
            id -> Int4,
            duration -> Nullable<Interval>,
        }
    }
}

#[derive(Queryable, Selectable, Insertable, Clone, Debug)]
#[diesel(table_name = schema::fantastic)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Fantastic {
    id: i32,
    #[diesel(
            deserialize_as = Option<PgInterval>,
            serialize_as = Option<PgInterval>
        )]
    duration: DurationProxy,
}

/// Represent a duration to deserialize from Postgres.
#[derive(Copy, Clone, Debug)]
pub struct DurationProxy(Option<Duration>);

impl From<Option<PgInterval>> for DurationProxy {
    fn from(value: Option<PgInterval>) -> Self {
        todo!()
    }
}

impl From<DurationProxy> for Option<PgInterval> {
    fn from(value: DurationProxy) -> Self {
        todo!()
    }
}
```

This code will fail to compile because `DurationProxy` is does not implement `Iterator` (or does not implement `.map(...)`).

Why would Diesel require `DurationProxy` to implement such methods? According to the [official documentation](https://docs.diesel.rs/master/diesel_derives/derive.Insertable.html#optional-field-attributes), I thought I have to implement `From<T>` and `Into<T>` to satisfy `deserialize_as=T` and `serialize_as=T` attributes.

**Comment by weiznich** [comment]
That's a very misleading error message from rustc. You should likely fill an upstream bug for this as `Iterator` is not mentioned (or required) anywhere here.

The underlying problem is that you've mixed up the annotations on your struct field. It needs to be:

```rust
    #[diesel(
            deserialize_as = DurationProxy,
            serialize_as = DurationProxy,
        )]
    duration: Option<PgInterval>,
```

+ the `DurationProxy` type needs to implement `FromSql` + `ToSql` as well (and derive `AsExpression` + `FromSqlRow`).

**Reply by McDic** [reply]
I think your suggested code is not correct. (If I am wrong, then sorry, please correct me.) If I understood correctly, following code means...

- We are using `SomeStruct::field_name` as T3 in business logic part
- When diesel deserialize `SomeStruct::field_name` by select query etc, it creates `T1` instance and then convert it to `T3` (`TryFrom<T1> for T3`) [(Official documentation of `Queryable`)](https://docs.diesel.rs/master/diesel_derives/derive.Queryable.html)
- When diesel serialize `SomeStruct::field_name` by insert query etc, it creates `T2` instance by converting `SomeStruct::field_name` (`From<T3> for T2`) [(Official documentation of `Insertable`)](https://docs.diesel.rs/master/diesel_derives/derive.Insertable.html)

```rust
#[derive(Queryable, Selectable, Insertable, Clone, Debug)]
struct SomeStruct {
    #[diesel(deserialize_as = T1, serialize_as = T2)]
    field_name: T3,
}
```

The following code from the [official documentation](https://docs.diesel.rs/master/diesel_derives/derive.Insertable.html#examples) shows such purpose.

```rust
// .. from the official documentation

#[derive(Insertable, PartialEq, Debug)]
#[diesel(table_name = users)]
struct InsertableUser {
    id: i32,
    #[diesel(serialize_as = UppercaseString)]
    name: String,
}

let user = InsertableUser {
    id: 1,
    name: "thomas".to_string(), // When this is saved on DB, 
                                // the entire string is converted to uppercase via `UppercaseString`,
                                // which is declared in `serialize_as=`
};
```

I wanted to use `DurationProxy` directly instead of `Option<PgInterval>` which is a pure data and does not provide any timedelta functionalities. The reason why I use `serialize_as=Option<PgInterval>` is because I don't have to implement `FromSql` and `ToSql` by my own, as `Option<PgInterval>` already implements it.

**Reply by weiznich** [reply, marked as answer]
Well in that case: As written above: The error message is not correct, there is no need to implement `Iterator` in any way. Feel free to fill a rustc bug report for that. Other than that: I do not have the capacity to fix your code here, so you are on your own.

The examples from the documentation work (they are tested in the CI), so that feature is known to work as expected.

**Reply by weiznich** [reply]
I had another look at the relevant derive code and it turns out  that this is a bug in the derive. It seems like nobody did try to use a `Option` as target type yet. 

I've filled https://github.com/diesel-rs/diesel/pull/4039 to fix that.
